src/conversions/either.rs

```diff
@@ -0,0 +1,146 @@
+#![cfg(feature = "either")]
+
+//! Conversion to/from
+//! [either](https://docs.rs/either/ "A library for easy idiomatic error handling and reporting in Rust applications")’s
+//! [`Either`] type to a union of two Python types.
+//!
+//! Use of a generic sum type like [either] is common when you want to either accept one of two possible
```

Comment:
Personally, I am not particularly fond of the `Either` type as I think it encourages type blindness, i.e. the two types usually have more specific roles than just `Left` and `Right` and defining an `enum` with `#[derive(FromPyObject)]` slapped onto it is usually the better approach IMHO.

Of course, the API surface of PyO3 might not be a suitable place to have this discussion. On the other hand, we to carefully consider the set of crates which we integrate directly and the list of potential candidates is unbounded.

Reply:
Thanks for the comment!

> Personally, I am not particularly fond of the Either type as I think it encourages type blindness, i.e. the two types usually have more specific roles than just Left and Right and defining an enum with #[derive(FromPyObject)] slapped onto it is usually the better approach IMHO.

I would generally advise that as well in the general case, except when your code gets littered by `IntOrString`, `SeriesOrFrame`, `ThisOrThat` types that are only used on a single line of code.  Sometimes you just want a once-off sum-type which doesn't exist in stdlib (`Result` aside, which has a specific use-case), either for taking it as an argument, or for returning it, or just for manual extraction. The sum-type of two variants would be perhaps the most common of all in type signatures and would cover lots of simple cases; for anything more complex you could always reach out for a hand-written object with a FromPyObject-derive.

Another difference from the general case is that when using pyo3, you're (normally) not writing a crate to be used and shared within Rust, with a proper Rust API surface and proper enums crafted for each particular case; your actual API surface lives in Python land which doesn't really care if it's `IntOrString` or `Either<i32, String>` after it's all compiled. (Anyways, that gets pretty subjective so it's my personal take on this.)

> Of course, the API surface of PyO3 might not be a suitable place to have this discussion. On the other hand, we to carefully consider the set of crates which we integrate directly and the list of potential candidates is unbounded.

A few points re: `either`, it's been 1.0 for around 7 years, is a tiny crate, has a single exported type, zero dependencies and tons of dependent crates (mostly used as a internal type rather than part of the public API, but still). I wouldn't expect it to affect the API surface that much or increase maintenance burden.

Reply:
> A few points re: either, it's been 1.0 for around 7 years, is a tiny crate, has a single exported type, zero dependencies and tons of dependent crates (mostly used as a internal type rather than part of the public API, but still). I wouldn't expect it to affect the API surface that much or increase maintenance burden.

Indeed, the maintenance effort seems to be bearable. I'd just like to avoid suggesting this approach to people, but we also do not put this on the frontend but it will be part of an ever growing list of integration features.

So let's put it like this: If this has any traction with the other maintainers, I consider my concerns overwritten as they are not severe enough to argue over for any amount of time.

Reply:
I have to agree with @adamreichold here, I've never wanted to use `either` myself as it's quite straightforward to derive most traits I want on a simple enum these days.

Similarly `either` is, as you say, very stable and it's unlikely we'll ever need to change this code other than for our own rearranging.

I'm probably a bit more lax than @adamreichold; I'd say that I myself don't want this but if there's multiple community members who want this then I can support it being merged.

Reply:
So what do we do here? There has been no more feedback (neither for or against) and both me and David are reserved but not completely opposed.

I'd say let's honour putting in the work and merge this.

Reply:
To that end, I've just rebased this PR.